eclair-core/src/main/scala/fr/acinq/eclair/io/IncomingConnectionsTracker.scala

```diff
@@ -0,0 +1,80 @@
+package fr.acinq.eclair.io
+
+import akka.actor.typed.delivery.DurableProducerQueue.TimestampMillis
+import akka.actor.typed.eventstream.EventStream
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+import akka.actor.typed.{ActorRef, Behavior}
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.NodeParams
+import fr.acinq.eclair.channel.ChannelOpened
+import fr.acinq.eclair.io.IncomingConnectionsTracker.Command
+import fr.acinq.eclair.io.Monitoring.Metrics
+import fr.acinq.eclair.io.Peer.Disconnect
+
+/**
+ * A singleton actor that limits the total number of incoming connections from peers that do not have channels with us.
+ *
+ * When a new incoming connection request is received, the Switchboard should send an
+ * [[IncomingConnectionsTracker.TrackIncomingConnection]] message.
+ *
+ * When the number of tracked peers exceeds `eclair.peer-connection.max-no-channels`, send [[Peer.Disconnect]] to
+ * the tracked peer with the oldest incoming connection.
+ *
+ * When a tracked peer disconnects or confirms a channel, we will stop tracking that peer.
+ *
+ * We do not need to track peers that disconnect because they will terminate if they have no channels.
+ * Likewise, peers with channels will disconnect and terminate when their last channel closes.
+ *
+ * Note: Peers on the sync whitelist are not tracked.
+
+ * This actor enables a DoS attack that can block new incoming liquidity from non-whitelisted nodes.
+ * An attacker can trivially disconnect other incoming connections before they can open channels by repeatedly opening
+ * new connections using random node IDs.
```

Comment:
I think it's worth being more explicit, "DoS attack" is scary (compared to what the attack actually is). It's also useful to mention the (future) fix, what about (also note the missing `*` in the line above):

```suggestion
 *
 * This rate-limiting can be abused by attackers to prevent us from accepting channels from unknown peers: attackers can
 * create a continuous stream of incoming connections with random nodeIds, which forces us to constantly disconnect old
 * connections before they have the opportunity to open a channel. This can be fixed by adding a TCP rate-limiter that
 * rejects connections based on IP addresses, which forces the attacker to own a lot of IP addresses.
```

Reply:
Good point; your suggestion is more clear/correct. Fixed in [9f70d19](https://github.com/ACINQ/eclair/pull/2601/commits/9f70d19d703f72d494daca0171ec6e7e63382737).